### PR TITLE
runtime: Support memory hotplug via virtio-mem on s390x

### PR DESCRIPTION
**Pull request by BbolroC**
Since QEMU v10.0.0 and Linux v6.13 support `virtio-mem-ccw`, this PR enables memory hotplug with virtio-mem on s390x.

Additional details are provided in the individual commit messages. The corresponding runtime-rs changes will be submitted separately.

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

### CODE REVIEW

File: src/runtime/virtcontainers/qemu_test.go [hunk 1]
```diff
@@ -825,3 +827,325 @@ func TestPrepareInitdataImage(t *testing.T) {
...
+		defer serverConn.Close()
+		hello := `{"QMP":{"version":{"qemu":{"micro":0,"minor":0,"major":5},"package":""},"capabilities":[]}}` + "\n"
+		if _, err := serverConn.Write([]byte(hello)); err != nil {
+			return
+		}
+		scanner := bufio.NewScanner(serverConn)
+		for _, resp := range responses {
+			if !scanner.Scan() {
+				return
+			}
```
[the remark this answers is not in the document]
Comment (BbolroC) [reply src/runtime/virtcontainers/qemu_test.go:847]: The server is designed just for mocking. The validation can be done in separate unit tests dealing with a QMP call.

File: src/runtime/virtcontainers/qemu_test.go [hunk 2]
```diff
@@ -825,3 +827,325 @@ func TestPrepareInitdataImage(t *testing.T) {
...
+	startTestQMPServer(t, serverConn, []string{`{"return":{}}`})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	disconnectedCh := make(chan struct{})
+	cfg := govmmQemu.QMPConfig{Logger: newQMPLogger()}
+	qmp, _, err := govmmQemu.QMPStartWithConn(ctx, clientConn, cfg, disconnectedCh)
+	assert.NoError(err)
+
```
[the remark this answers is not in the document]
Comment (BbolroC) [reply src/runtime/virtcontainers/qemu_test.go:874]: That is too much for this PR.

**Comment by BbolroC** [comment]
I've switched it to a draft because 2 tests failures for the k8s integration test, which I didn't expected, were observed. Thanks!

**Comment by BbolroC** [comment]
The PR was tested on IBM SEL successfully. Thanks!

File: src/runtime/virtcontainers/qemu.go [hunk 3]
```diff
@@ -2243,7 +2272,27 @@ func (q *qemu) hotplugMemory(memDev *MemoryDevice, op Operation) (int, error) {
 
 }
 
+// resizeVirtioMem resizes the virtio-mem device to the specified size in MB
+func (q *qemu) resizeVirtioMem(newSizeMB int) error {
+	sizeByte := uint64(newSizeMB) * 1024 * 1024
+	err := q.qmpMonitorCh.qmp.ExecQomSet(q.qmpMonitorCh.ctx, "virtiomem0", "requested-size", sizeByte)
+	if err != nil {
+		q.Logger().WithError(err).Error("failed to resize virtio-mem device")
```
[the remark this answers is not in the document]
Comment (Apokleos) [reply src/runtime/virtcontainers/qemu.go:2285]: I think the Copilot's suggestion is fine. `converts newSizeMB to uint64`

File: src/runtime/virtcontainers/qemu.go [hunk 4]
```diff
@@ -2458,13 +2481,10 @@ func (q *qemu) ResizeMemory(ctx context.Context, reqMemMB uint32, memoryBlockSiz
 	var addMemDevice MemoryDevice
 	if q.config.VirtioMem && currentMemory != reqMemMB {
 		q.Logger().WithField("hotplug", "memory").Debugf("resize memory from %dMB to %dMB", currentMemory, reqMemMB)
-		sizeByte := uint64(reqMemMB - q.config.MemorySize)
-		sizeByte = sizeByte * 1024 * 1024
-		err := q.qmpMonitorCh.qmp.ExecQomSet(q.qmpMonitorCh.ctx, "virtiomem0", "requested-size", sizeByte)
-		if err != nil {
+		newSizeMB := int(reqMemMB) - int(q.config.MemorySize)
```
Comment (Apokleos) [inline src/runtime/virtcontainers/qemu.go:2515]: If we adopt the Copilot's suggestion on the resizeVirtioMem with a parameter of `newSizeMB uint64`, it seems that we can keep the L2461~L2462 as they were. As we should also check for negative value of the   `newSizeMB` based on your current changes.

Comment (BbolroC) [reply src/runtime/virtcontainers/qemu.go:2515]: Yep, you are right, but in resizeVirtioMem():

```
func (q *qemu) resizeVirtioMem(newSizeMB int) error {
	if newSizeMB < 0 {
		return fmt.Errorf("cannot resize virtio-mem device to negative size (%d) memory", newSizeMB)
	}
    ...
}
```
Any calculation result ending up with a negative value should be caught by this condition. At this time, I would like to reject the Copilot's suggestion. Thanks!

Comment (stevenhorsman) [reply src/runtime/virtcontainers/qemu.go:2515]: Is it worth adding a unit test to demonstrate this error in the negative case?

File: src/runtime/virtcontainers/qemu.go [hunk 5]
```diff
@@ -995,37 +995,64 @@ func (q *qemu) setupVirtioMem(ctx context.Context) error {
...
+		}
+
+		defer func() {
+			if err != nil {
+				q.arch.removeDeviceFromBridge("virtiomem-dev")
+			}
+		}()
+
+		devAddr = addr
+		bus = bridge.ID
```
Comment (stevenhorsman) [inline src/runtime/virtcontainers/qemu.go:1036]: What was the reason for changing `bridgeID := bridge.ID` to this and defining bus above. I don't think it's being used before this point?

Comment (BbolroC) [reply src/runtime/virtcontainers/qemu.go:1036]: The change was made only to match the last function parameter name `bus` in `ExecMemdevAdd()`, making the code more readable and consistent. If you are not comfortable with this. I will get it back to the original again. Thanks!

**Comment by BbolroC** [comment]
> Is this functionality ported/going to be ported to runtime-rs?

Yep, it will be a task for one of my colleagues. 😉

File: src/runtime/virtcontainers/qemu.go (hunk 4 above)
Comment (BbolroC) [reply src/runtime/virtcontainers/qemu.go:2515]: > Is it worth adding a unit test to demonstrate this error in the negative case?

Yep, I will add it. Thanks!

File: src/runtime/virtcontainers/qemu.go (hunk 5 above)
Comment (stevenhorsman) [reply src/runtime/virtcontainers/qemu.go:1036]: I'm ok either way, I just noticed the diff, so thought I'd ask.
